### PR TITLE
[IMP] core: allow direct sql computations for fields

### PR DESCRIPTION
**Pull request by yelizariev**
Comparision with regular computed fields:
1. it work faster
2. allows read_group without making field storable

---

task-1915411

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by rco-odoo — COMMENTED** [review]
Things are slowly getting clearer to me.

So the query function (or whatever we eventually call it) is a way to get an SQL expression for the field.  It basically provides something to replace `f'"{alias}"."{field.name}"'` in any SQL query: in a SELECT clause (`read`, `read_group`), in a WHERE clause (domain), in an ORDER BY clause, or even in a GROUP BY clause (`read_group`).

Some questions remain:
* You cannot inject parameters in the SQL expression, is this limitation acceptable?
* Would it make sense to combine `compute` with an SQL expression (to be able to GROUP BY a non-stored computed field, for instance) ?
* Unlike most ORM features, there is no built-in security in this feature.  Isn't it too risky?
* Are average developers capable of managing table aliases and joins properly?  Isn't it too complex?

The last point is important to me.  Not all developers are smart, and even smart developers are not always smart.  Wouldn't this feature make it too easy to shoot yourself in the foot?  Wouldn't it generate a plethora of SQL injections because of its limitations?

File: odoo/addons/test_new_api/models/test_new_api.py [hunk 1]
```diff
@@ -1277,3 +1278,64 @@ class TriggerRight(models.Model):
...
+            WHEN "{table}".deadline - (CURRENT_DATE AT TIME ZONE "{table}".tz)::date = 0 THEN 'today'
+            ELSE null
+        END""".format(table=table_alias)
+
+    @api.model
+    @api.depends('partner_id.name')
+    def _query_partner_name(self, table_alias, query):
+        field = 'name'
+        alias = query.left_join(table_alias, 'id', 'res_partner', 'id', field)
+        return '"%s"."%s"' % (alias, field)
```
[the remark this answers is not in the document]
Comment (yelizariev) [reply odoo/addons/test_new_api/models/test_new_api.py:1319]: Yeah, I doubted about it. I took it from ``_generate_translated_field``

https://github.com/odoo/odoo/blob/d583de02392b7738e7eead4ef33f91ef8a47b0f6/odoo/models.py#L4293-L4294

File: odoo/addons/test_new_api/models/test_new_api.py [hunk 2]
```diff
@@ -1277,3 +1278,64 @@ class TriggerRight(models.Model):
...
+            ('today', 'Today'),
+            ('planned', 'Planned')
+        ],
+        'State',
+        query='_query_state'
+    )
+    deadline = fields.Date('Due Date')
+    tz = fields.Char('Timezone', default='UTC')
+    partner_id = fields.Many2one('res.partner')
+    partner_name = fields.Char(query='_query_partner_name')
```
[the remark this answers is not in the document]
Comment (yelizariev) [reply odoo/addons/test_new_api/models/test_new_api.py:1299]:
```
    partner_name = fields.Char(sql='_sql_partner_name')
    partner_name = fields.Char(select='_select_partner_name')
```
?

**Comment by william-andre** [comment]
I would really love this for related fields too

**Comment by yelizariev** [comment]
I have kind of the same concerns, but we already give access to ``cr.execute``... The new feature is kind of simplified version of ``cr.execute``, i.e. you specify minimal sql stuff and then use ORM methods.

> * You cannot inject parameters in the SQL expression, is this limitation acceptable?

I'm not sure that I understand you. Do you have an example? Can't we just read System Parameters?

```{alias}.qty * (select value from ir_config_parameter where name='price')::numeric```

> * Would it make sense to combine `compute` with an SQL expression (to be able to GROUP BY a non-stored computed field, for instance) ?

My understanding is that new feature will be used in rare cases for optimization. But once sql expression is set, ``compute`` attribute cannot be used.

> * Unlike most ORM features, there is no built-in security in this feature.  Isn't it too risky?

Existing security restrictions on model level (ir.rule, ir.model.access) are not enough? But yes, it's important question. We need to think carefully about it.

> * Are average developers capable of managing table aliases and joins properly?  Isn't it too complex?

Yes, it could be complex. We may need to document ``query`` methods (left_join, etc.). But again, ``cr.execute`` is also complex and risky

> The last point is important to me. Not all developers are smart, and even smart developers are not always smart. Wouldn't this feature make it too easy to shoot yourself in the foot? Wouldn't it generate a plethora of SQL injections because of its limitations?

Maybe we should block using of ``self.env.context``, recordset values (if they are available) to avoid sql injections if a developer pass user input without proper escaping. But again, it much less risky, than having access to ``cr.execute``

**Comment by yelizariev** [comment]
> I would really love this for related fields too

I think we could generate SQL Expression for related fields automatically and reuse it in read_group method

**Comment by william-andre** [comment]
I had an idea of implementation for this https://discord.com/channels/678381219515465750/687339689522364423/738015758562623489

We can add the computed fields inside of it too.




________

Use a SQL VIEW for models on top of the TABLE.
If we have a model with
```python
class MyModel(models.Model):
    _name = 'my.model'
   
    own_field = fields.Many2one()
    related_field = fields.Integer(related='own_field.other')
```

It would have one table and one view.
* The table would only store `own_field` (+ various ORM fields)
* The view would be something like 
  ```sql
  SELECT table.*,
         related_table.other AS related_field
  FROM table 
  JOIN related_table ON related_table.id = table.own_field
  ```
* All the queries would be done on the VIEWs, the TABLEs are only there internally

This would
* reduce the space of the db (quick search tells me 156+ fields in Odoo: regex \(.*related=[^)]*store=True  )
* have all related fields accessible from SQL queries without have to set `store=True`
* make https://www.odoo.com/web#id=2155109&action=333&active_id=1364&model=project.task&view_type=form&cids=1&menu_id=4720 a trivial fix
  We can also use related fields in _order, group_by, and other SQL depending params
* make it possible to change values that are related on a lot of records (i.e. changing the type of an account without having to update all the account.move.line related to that account) 
* if some JOINs are complex to do with osv.query.Query (resulting in huge lists of ids), just declare the related and it is done by the SQL server. https://www.odoo.com/web?#id=2238691&action=333&active_id=1364&model=project.task&view_type=form&cids=1&menu_id=4720
* Maybe simplify some code in Odoo because you can UPDATE a simple view https://www.postgresql.org/docs/12/rules-views.html#RULES-VIEWS-UPDATE

**Comment by yelizariev** [comment]
Before deleting ``store=True`` for related fields, we have to check that the field is not used via raw sql request.

Using views would make sql request clearer, but it seems much harder to implement

**Comment by odony** [comment]
As the saying goes: use case or die :p

Without a good series of clear cut use cases where you can demonstrate that this features provides a net benefit, this looks like the antithesis of what we want for the framework, and a solution looking for a problem, too :D

Reminder: what we want for the framework:

1) Simplifying our API.
This makes the API more complex but adding a super-advanced feature that most developers do NOT need, but will be tempted to use anyway, for the wrong reasons. And will have trouble to make it work. `alias`? `query`? WTF? Developers should not need to understand our query builder internals to implement a simple model. Yet it's impossible to implement anything beyond trivial without understanding how your `sql` will be used in the query builder.

2) Avoiding multiple different ways to do the same thing.
Here we're adding a non-obvious way for implementing non-stored computed fields, and now developers are going to have to choose if they want a normal computed field or a "raw sql" one. And they're going to choose the wrong option, because it's super hard to know.

3) Keeping it as modular as possible
Hardcoding the field implementation in pure SQL means that any override of the logic must either perform awful and dangerous SQL text manipulation, or completely discard the parent implementation. Both are guaranteed to break things *very* often. Imagine that you want to add the management of "discount" in your `_sql_total` example...


**So: use cases?**

Please provide actual examples in the current codebase where:
- the current field implementation is done correctly
- yet an `sql` based implementation would be:
  - significantly faster (measure it) without becoming unreadable ; or
  - significantly more readable without becoming slower; and
  - not completely impossible to extend in a modular fashion

My gut feeling is that apart from some very rare oddities, all of them will more elegantly solved with a stored field, or a non-stored computed field with a `search` function.

Stored fields are not evil, and denormalization is a key part of database optimization! The last proposition by @william-andre is a good idea to completely destroy performance :P

------

PS: the examples in the PR demonstrate cases where using raw SQL is a bad idea in general:
- Timezone manipulation: in many cases rely on the user's timezone, which it not part of the current table (and joining res_users -> res_partner would be a perf killer). That is why the ORM has high-level primitives for cases where working in pure UTC is not enough. Your example is also purposely(?) skewed: we either use datetimes and UTC as reference, or we used pure client-relative dates and no timezones, but mixing them is a usability disaster that makes no sense.
- Float/monetary manipulation: difficult to do correctly in SQL (and regularly broken when devs think they know how to do it). Relies on arbitrary precisions and roundings that come from other master data. All handled properly by high-level ORM features.
- Sub-select on lines: reimplements the ORM's o2m, and is a clear cut case with the cost of the operation justifies a stored field (+ rounding issues again)

File: odoo/addons/test_new_api/models/test_new_api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xmo-odoo) [reply odoo/addons/test_new_api/models/test_new_api.py]: `sql` seems good.

**Comment by yelizariev** [comment]
Use cases are
###  read_progress_bar 

https://www.odoo.com/web#cids=1&id=1915411&model=project.task&view_type=form&menu_id=

currently opening Dashboard in [some](https://www.odoo.com/web#cids=1&id=2346901&menu_id=4720&model=project.task&view_type=form) customer database takes 10 seconds. Could be dropped to less than a second, I suppose. I could measure.

### read_group on non-stored related fields 

https://www.odoo.com/web#id=2155109&action=4043&model=project.task&view_type=form&cids=1&menu_id=4720

currently you have to add stored related field every time you want to make grouping to get some statistic https://docs.google.com/presentation/d/1eAn7tmjR_X3CC_t6cZVJ7r8wBnSSubkBqtKgCXvLAQo/edit#slide=id.g6be6d36f1c_0_6

-------

I kind of agree about *what we want for the framework*. A solution could be making the same thing behind the scene and don't provide direct access to *sql* methods:

###  related fields

``_sql_partner_name`` could be generated from related field definition, or even without adding new field, but just in ``read_group`` method if ``groupby`` contains joined fields (e.g. ``partner_id.country_id.name``)

### computed field

Idea is to try generate an *sql* method out of compute function. If it's  not possible, then stay with usual process.

For example,
```
    @api.depends_context('partner_id')
    def _compute_product_code(self):
        for product in self:
            for supplier_info in product.seller_ids:
                if supplier_info.name.id == product._context.get('partner_id'):
                    product.code = supplier_info.product_code or product.default_code
                    break
            else:
                product.code = product.default_code
```

To generate *sql* method, we could first check if it has only simple for-loop. If yes, then extract for-body:

```
            for supplier_info in product.seller_ids:
                if supplier_info.name.id == product._context.get('partner_id'):
                    product.code = supplier_info.product_code or product.default_code
                    break
            else:
                product.code = product.default_code
```

Then eval it with ``product`` equal to a special object, that can accumulate operations made on it and convert it to sql. Kind of [query builder](https://github.com/kayak/pypika):
> image

It's a more simple example than [computing state depending on timezone and current time](https://github.com/odoo/odoo/blob/baebf5814425cce6a9481b4efec5248bbba3ea64/addons/mail/models/mail_activity.py#L203-L226), but I think something similar could be done in that case too.

**Comment by fpodoo** [comment]
Can't agree more with @odony 

Why would someone do SQL fields when you can do store=True, that are faster on read, easier to implement, and have less side effects / risks. read_progress_bar is a bad example: the solution would be to use stored fields; it's faster than SQL fields as these fields are way more often read than written.

The simpler the framework, the easier it is to optimize it for performance. So, we shouldn't complexify it unless we have a very good reason.

Instead of trying to find use cases to justify this solution, what about cancelling this PR? (if one day, we have enough real use cases that justifies it, we can reuse the PR, but it does not look like we have clear ones now.)

**Comment by william-andre** [comment]
> Stored fields are not evil, and denormalization is a key part of database optimization! The last proposition by @william-andre is a good idea to completely destroy performance :P

Can you elaborate please? I have made a POC here https://github.com/odoo-dev/odoo/commit/03ea0b69d82a5ce3763f1823f057892378efbfb0

```
psql (12.6 (Ubuntu 12.6-0ubuntu0.20.04.1))
Type "help" for help.

master-related-view-wan=> EXPLAIN ANALYZE SELECT name FROM account_move;
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Seq Scan on account_move  (cost=0.00..3.14 rows=14 width=16) (actual time=0.034..0.041 rows=14 loops=1)
 Planning Time: 1.237 ms
 Execution Time: 0.128 ms
(3 rows)

master-related-view-wan=> EXPLAIN ANALYZE SELECT name FROM account_move_related;
                                                  QUERY PLAN                                                  
--------------------------------------------------------------------------------------------------------------
 Seq Scan on account_move move  (cost=0.00..3.14 rows=14 width=16) (actual time=0.015..0.022 rows=14 loops=1)
 Planning Time: 4.172 ms
 Execution Time: 0.080 ms
(3 rows)

master-related-view-wan=> EXPLAIN ANALYZE SELECT company_id FROM account_move_related;
                                                 QUERY PLAN                                                  
-------------------------------------------------------------------------------------------------------------
 Seq Scan on account_move move  (cost=0.00..3.14 rows=14 width=4) (actual time=0.016..0.026 rows=14 loops=1)
 Planning Time: 0.939 ms
 Execution Time: 0.056 ms
(3 rows)

master-related-view-wan=> EXPLAIN ANALYZE SELECT company_id_rel FROM account_move_related;
                                                        QUERY PLAN                                                        
--------------------------------------------------------------------------------------------------------------------------
 Hash Right Join  (cost=3.31..17.92 rows=14 width=4) (actual time=0.115..0.137 rows=14 loops=1)
   Hash Cond: (journal.id = move.journal_id)
   ->  Seq Scan on account_journal journal  (cost=0.00..12.10 rows=210 width=8) (actual time=0.016..0.018 rows=7 loops=1)
   ->  Hash  (cost=3.14..3.14 rows=14 width=8) (actual time=0.047..0.048 rows=14 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 9kB
         ->  Seq Scan on account_move move  (cost=0.00..3.14 rows=14 width=8) (actual time=0.017..0.029 rows=14 loops=1)
 Planning Time: 1.245 ms
 Execution Time: 0.194 ms
(8 rows)
```
By using `LEFT JOIN` (or `OUTER`, more generally), the query plan to read non related fields is exactly the same. You will need to actually do the join only when you really need it: when you want to use a `GROUP BY`, or use a `LIMIT`, or anything that is not doable in the ORM because we do not want to fetch everything.

Use cases are of course to use those fields in the pivot/graph/BI views, or also in the web client search views.

Of course, since `company_id` is critical it would be useful to store it; but is is not the case for each field.


_______

The ORM could do the JOIN without using this implementation; and it would allow to not declare the field at all. For instance, we could have made this instead of that https://github.com/odoo/odoo/pull/66469/files

```python
    def _compute_pos_order_amount_total(self):
        self_pos = self.filtered(lambda t: t.team_type == 'pos')
        res = self.env['report.pos.order'].read_group(
            [("config_id.crm_team_id", "in", self_pos.ids), ("session_id.state", "=", "opened")],
            ["price_total"],
            ["config_id.crm_team_id"],
        )
        res_dict = {r['config_id__crm_team_id']: r['price_total'] for r in res}
        for team in self.self_pos:
            team.pos_order_amount_total = res_dict(team.id)
```

**Comment by yelizariev** [comment]
> Why would someone do SQL fields when you can do store=True, that are faster on read, easier to implement, and have less side effects / risks. read_progress_bar is a bad example: the solution would be to use stored fields; it's faster than SQL fields as these fields are way more often read than written.

You cannot use stored fields for read_progress_bar until you recompute it by cron every day

**Comment by william-andre** [comment]
A real feature that has been asked a lot is to be able to group partners by followup level. It is impossible to store that field without using a cron.
There is a search function, but I have been asked many times to implement the group by; which the ORM is not capable to do because it has no idea how to do it in the database.

**Comment by pedrobaeza** [comment]
Maybe the solution is that the framework provide a `groupby` argument to the field in the same sense of `search` for providing methods?

For current framework, I have hacked group by in some modules interfering the `read_group`, like in https://github.com/OCA/hr/blob/b4401a83229f4f2b49806e9ec8774db2d95ede5f/hr_attendance_report_theoretical_time/reports/hr_attendance_theoretical_time_report.py#L215

**Comment by william-andre** [comment]
https://discord.com/channels/678381219515465750/687339689522364423/813421768953167912
Funny how there are 2 questions in the `rd-framework-py` channel that are questioning the system with the stored related at the same time.

* A stored related added with studio is not populated like it should in some cases
* Ask if there is a way to do a group by on a non stored related

We are not making this up... Maybe we don't agree on the best solution but there is a problem here with the current solution and I think it is wrong to simply discard it.

**Comment by yelizariev** [comment]
@william-andre in few days I'll finish a new implementation which:
* doesn't change fields API
* cleans and speeds up read_progress_bar
* allows grouping by non-storable related fields

stay tuned :relaxed:

**Comment by rco-odoo** [comment]
@william-andre your example does not disqualify @odony 's remark.  Olivier warns about the API being too obscure for most developers, which I pretty much agree with.  I suggested this API, because that's the only way IMHO to add tables and/or joins to the current query and deal with table aliases (which are used when searching/read_grouping).  But it really is a low-level ORM implementation trick.

I noticed that most fields being asked for are *related fields*.  We could provide some built-in support for related fields in `read_group`, and it would do the job.  The only tricky part is access rights when the field is not `compute_sudo=True` (it is by default, though.)

**Comment by william-andre** [comment]
I would indeed already be really happy with any solution that would make the `read_group` work without storing the related fields (or even without declaring the related field).
I think it wouldn't change anything for the non-metastorm devs but it would all to do a lot of useful things. :+1:

**Pull request closed without merging**